runtime/moonbeam/src/migrations.rs

```diff
@@ -17,27 +17,303 @@
+		// compute the hash of the expected status
+		let mut expected_hash: Vec<u8> = Vec::new();
+		let mut actual_hash: Vec<u8> = Vec::new();
+		// track reads and writes
+		let mut num_reads = 0;
+		let mut num_writes = 0;
+
+		// compute the complete hash of the expected state (the corrupted one)
+		// compute the complete hash of the current state
+		// if they are the same, it means tha the migration must be executed
+		// NOTE: the referenda in the corrupted state are long closed so they cannot be updated anymore,
+		// if the migration runs successfully once, the state will never match the corrupted state,
+		// this ensures idempotence of the changes
+		for (ref_index, ref_storage_key, wrong_storage_value, correct_storage_value) in
+			CORRUPTED_REFERENDA
+		{
+			let k = from_hex(ref_storage_key);
+			if k.is_err() {
+				log::error!("cannot deserialize storage key {:?}", k.err());
+				return <Runtime as frame_system::Config>::DbWeight::get().reads(num_reads);
+			}
+			let v = from_hex(wrong_storage_value);
+			if v.is_err() {
+				log::error!("cannot deserialize storage value (1) {:?}", v.err());
+				return <Runtime as frame_system::Config>::DbWeight::get().reads(num_reads);
+			}
+			// validate also the data that will be written later
+			let v1 = from_hex(correct_storage_value);
+			if v1.is_err() {
+				log::error!("cannot deserialize storage value (2) {:?}", v1.err());
+				return <Runtime as frame_system::Config>::DbWeight::get().reads(num_reads);
+			}
+
+			let key = k.unwrap();
+			let value = v.unwrap();
+
+			let data = [expected_hash, key.clone(), value].concat();
+			expected_hash = frame_support::Blake2_256::hash(&data).to_vec();
+
+			let current_storage = get_raw(&key);
+			num_reads += 1;
+			if current_storage.is_none() {
+				log::error!("storage for referenda {:?} not found", ref_index);
+				return <Runtime as frame_system::Config>::DbWeight::get().reads(num_reads);
+			}
+
+			let data = [actual_hash, key, current_storage.unwrap()].concat();
+			actual_hash = frame_support::Blake2_256::hash(&data).to_vec();
+		}
+
+		if expected_hash != actual_hash {
+			// the migration was already executed, track the reads
+			log::warn!(
+				"expected partial state hash {:?} doesn't match actual {:x?}, \
+				has the migration been already applied?",
+				sp_core::hexdisplay::HexDisplay::from(&expected_hash),
+				sp_core::hexdisplay::HexDisplay::from(&actual_hash),
+			);
+			return <Runtime as frame_system::Config>::DbWeight::get().reads(num_reads);
+		}
+
+		log::info!(
+			"matched expected partial state hash {:?} with actual {:x?}",
+			sp_core::hexdisplay::HexDisplay::from(&expected_hash),
+			sp_core::hexdisplay::HexDisplay::from(&actual_hash),
+		);
+
+		// update the storage
+		for (ref_index, ref_storage_key, _, correct_storage_value) in CORRUPTED_REFERENDA {
+			// we can safely unwrap as we already check for correctness before
+			let k = from_hex(ref_storage_key).unwrap();
```

Comment:
We should never use unwrap() inside runtime production code, please handle errors instead

Reply:
the errors are handled in the previous loop https://github.com/moonbeam-foundation/moonbeam/blob/cf48f6919fb93bd2813b727e06b8cb7d92c9a040/runtime/moonbeam/src/migrations.rs#L201 where all the keys and values are checked for correctness before executing the update loop.

Reply:
I strongly recommend to use hex_literal instead: https://docs.rs/hex-literal/latest/hex_literal/
It's a crate that provide a macro to decode hexa strings at compile time